### ISSUE TITLE
Multicast is slower than expected in fastdp

### ISSUE THREAD

**Issue by bboreham** [weaveworks/weave#2003]
Because the receiving weaver is getting called on every packet via the ODP miss handler

Testing with iperf, in debug mode we see lots of these:

 ````
DEBU: 2016/02/28 17:50:43.307011 ODP miss map[20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 3:InPortFlowKey{vport: 1} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000101, mask: ffffffffffffffffffffffff} 10:BlobFlowKey{type: 10, key: d0b51389, mask: ffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff}] on port 1
DEBU: 2016/02/28 17:50:43.307245 ODP miss map[15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 3:InPortFlowKey{vport: 1} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000102, mask: ffffffffffffffffffffffff} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff}] on port 1
DEBU: 2016/02/28 17:50:43.307446 ODP miss map[6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000101, mask: ffffffffffffffffffffffff} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 10:BlobFlowKey{type: 10, key: d0b51389, mask: ffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 3:InPortFlowKey{vport: 1}] on port 1
DEBU: 2016/02/28 17:50:43.307493 ODP miss map[16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 3:InPortFlowKey{vport: 1} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000102, mask: ffffffffffffffffffffffff}] on port 1
DEBU: 2016/02/28 17:50:43.307493 ODP miss map[4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 3:InPortFlowKey{vport: 1} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 10:BlobFlowKey{type: 10, key: d0b51389, mask: ffffffff} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000101, mask: ffffffffffffffffffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff}] on port 1
DEBU: 2016/02/28 17:50:43.308422 ODP miss map[16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000102, mask: ffffffffffffffffffffffff} 3:InPortFlowKey{vport: 1} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff}] on port 1
DEBU: 2016/02/28 17:50:43.308549 ODP miss map[16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000101, mask: ffffffffffffffffffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 3:InPortFlowKey{vport: 1} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 10:BlobFlowKey{type: 10, key: d0b51389, mask: ffffffff}] on port 1
DEBU: 2016/02/28 17:50:43.308634 ODP miss map[19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 3:InPortFlowKey{vport: 1} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000102, mask: ffffffffffffffffffffffff} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff}] on port 1
DEBU: 2016/02/28 17:50:43.308738 ODP miss map[10:BlobFlowKey{type: 10, key: d0b51389, mask: ffffffff} 3:InPortFlowKey{vport: 1} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 7:BlobFlowKey{type: 7, key: 0a200001e001020311000101, mask: ffffffffffffffffffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff}] on port 1
DEBU: 2016/02/28 17:50:43.308759 ODP miss map[7:BlobFlowKey{type: 7, key: 0a200001e001020311000102, mask: ffffffffffffffffffffffff} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 3:InPortFlowKey{vport: 1} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff}] on port 1
````

(they look very different because map iteration is randomised; here are the first two and last two with contents sorted:

 ````
DEBU: 2016/02/28 17:50:43.307011 ODP miss map[
2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 
3:InPortFlowKey{vport: 1} 
4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 
6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 
7:BlobFlowKey{type: 7, key: 0a200001e001020311000101, mask: ffffffffffffffffffffffff} 
10:BlobFlowKey{type: 10, key: d0b51389, mask: ffffffff} 
15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 
16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 
19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff}
20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 
] on port 1
DEBU: 2016/02/28 17:50:43.307245 ODP miss map[
2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff}
4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 
6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 
7:BlobFlowKey{type: 7, key: 0a200001e001020311000102, mask: ffffffffffffffffffffffff} 
15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 
16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 3:InPortFlowKey{vport: 1} 
19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 
20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 
] on port 1
...
DEBU: 2016/02/28 17:50:43.308738 ODP miss map[
3:InPortFlowKey{vport: 1} 
4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 
6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 
7:BlobFlowKey{type: 7, key: 0a200001e001020311000101, mask: ffffffffffffffffffffffff} 
10:BlobFlowKey{type: 10, key: d0b51389, mask: ffffffff} 
15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff}
16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 
19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 
20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 
] on port 1
DEBU: 2016/02/28 17:50:43.308759 ODP miss map[
2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 
3:InPortFlowKey{vport: 1} 
4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: 01:00:5e:01:02:03} 
6:BlobFlowKey{type: 6, key: 0800, mask: ffff}
7:BlobFlowKey{type: 7, key: 0a200001e001020311000102, mask: ffffffffffffffffffffffff} 
15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 
16:TunnelFlowKey{id: 0000000000065bb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 50017, tpdst: 6784} 
19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 
20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 
] on port 1

````

**Comment by awh** [comment]
@bboreham is the peer in question involved in any sleeve connections?

**Comment by bboreham** [comment]
In my tests there are just two peers created by the `weaveworks/weave/test` vagrant file. There is no reason for any Sleeve. 

From what I've been able to pick out, it creates a rule to send the packet ton port 0, which is the 'internal' vport. 

At the very least, creating the same rule every time is wrong.

**Comment by bboreham** [comment]
Oh, I tried it with the 1.2-style fastdp datapath only, and with the 1.4-style extra bridge, and they both go slowly. I've been debugging the former since fewer moving parts.

**Comment by bboreham** [comment]
Having extended the logging even more, here is a log from a `ping -b 10.47.255.255`, receiving side:

You can see the 'Creating new flow' line is repeated (first one is different, but after that).

@awh there is mention of sleeve heartbeats; is that relevant?

@dpw can you comment on what is supposed to happen after the first broadcast/multicast packet?

````
INFO: 2016/02/29 02:44:33.973758 Command line options: map[dns-effective-listen-address:172.17.0.1 http-addr:127.0.0.1:6784 ipalloc-range:10.32.0.0/12 nickname:host1 port:6783 datapath:weave log-level:debug name:1a:1e:21:54:f9:75 dns-listen-address:172.17.0.1:53]
INFO: 2016/02/29 02:44:33.973882 Command line peers: [192.168.48.12]
DEBU: 2016/02/29 02:44:33.974260 makeBridgeVport: {0 {{weave} 2 internal}}internal
DEBU: 2016/02/29 02:44:33.974300 makeBridgeVport: {1 {{vxlan-6784} 6784}}vxlan
INFO: 2016/02/29 02:44:33.974328 Communication between peers is unencrypted.
INFO: 2016/02/29 02:44:34.039948 Our name is 1a:1e:21:54:f9:75(host1)
INFO: 2016/02/29 02:44:34.045731 Docker API on unix:///var/run/docker.sock: &[BuildTime=2016-02-22T21:40:35.421962772+00:00 Version=1.10.2 ApiVersion=1.22 GitCommit=c3959b1 GoVersion=go1.5.3 Os=linux Arch=amd64 KernelVersion=4.2.0-30-generic]
INFO: 2016/02/29 02:44:34.045772 Assuming quorum size of 2
INFO: 2016/02/29 02:44:34.045833 [allocator 1a:1e:21:54:f9:75] Deleting persisted data for peername 00:00:00:00:00:00
INFO: 2016/02/29 02:44:34.071414 Listening for DNS queries on 172.17.0.1
INFO: 2016/02/29 02:44:34.071536 Sniffing traffic on weave (via ODP)
2016/02/29 02:44:34 ->[192.168.48.12:6783] attempting connection
INFO: 2016/02/29 02:44:34.097792 Listening for HTTP control messages on 127.0.0.1:6784
2016/02/29 02:44:34 ->[192.168.48.12:6783|82:43:51:a8:79:d4(host2)]: connection ready; using protocol version 2
INFO: 2016/02/29 02:44:34.115404 overlay_switch ->[82:43:51:a8:79:d4(host2)] using fastdp
2016/02/29 02:44:34 ->[192.168.48.12:6783|82:43:51:a8:79:d4(host2)]: connection added (new peer)
DEBU: 2016/02/29 02:44:34.120318 InvalidateRoutes
DEBU: 2016/02/29 02:44:34.120410 fastdp ->[192.168.48.12:6784|82:43:51:a8:79:d4(host2)]: confirmed
DEBU: 2016/02/29 02:44:34.120450 sleeve ->[192.168.48.12:6783|82:43:51:a8:79:d4(host2)]: Confirm
DEBU: 2016/02/29 02:44:34.120496 sleeve ->[192.168.48.12:6783|82:43:51:a8:79:d4(host2)]: confirmed
DEBU: 2016/02/29 02:44:34.120513 sleeve ->[192.168.48.12:6783|82:43:51:a8:79:d4(host2)]: sendHeartbeat
DEBU: 2016/02/29 02:44:34.121169 InvalidateRoutes
DEBU: 2016/02/29 02:44:34.121378 [allocator 1a:1e:21:54:f9:75]: Allocator.OnGossip: 659 bytes
DEBU: 2016/02/29 02:44:34.124415 fastdp ->[192.168.48.12:6784|82:43:51:a8:79:d4(host2)]: sendHeartbeat
DEBU: 2016/02/29 02:44:34.124956 sleeve ->[192.168.48.12:6783|82:43:51:a8:79:d4(host2)]: handleHeartbeat
...
DEBU: 2016/02/29 02:46:34.151404 fastdp ->[192.168.48.12:6784|82:43:51:a8:79:d4(host2)]: sendHeartbeat
DEBU: 2016/02/29 02:46:34.151598 sleeve ->[192.168.48.12:6783|82:43:51:a8:79:d4(host2)]: sendHeartbeat
DEBU: 2016/02/29 02:46:34.184093 sleeve ->[192.168.48.12:6783|82:43:51:a8:79:d4(host2)]: handleHeartbeat
DEBU: 2016/02/29 02:46:34.192087 ODP miss map[20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 16:TunnelFlowKey{id: 000000000084dbb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 41402, tpdst: 6784} 3:InPortFlowKey{vport: 1} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 4:EthernetFlowKey{src: 00:00:00:00:00:00, dst: 00:00:00:00:00:00}] on port 1
DEBU: 2016/02/29 02:46:34.192194 getMissHandler: 1 0x57c9a0
DEBU: 2016/02/29 02:46:34.192234 vxlan missHandler: 0x57e8e0 82:43:51:a8:79:d4(host2) 1a:1e:21:54:f9:75(host1)
DEBU: 2016/02/29 02:46:34.192288 Creating new flow:
MultiFlowOp: false [
 vxlanSpecialPacketFlowOp: src 82:43:51:a8:79:d4(host2) sender 192.168.48.12:6784, odpFlowKeyFlowOp: InPortFlowKey{vport: 1},
]
DEBU: 2016/02/29 02:46:34.192319 fastdp.send: vxlanSpecialPacketFlowOp: src 82:43:51:a8:79:d4(host2) sender 192.168.48.12:6784
DEBU: 2016/02/29 02:46:34.192349 fastdp ->[192.168.48.12:6784|82:43:51:a8:79:d4(host2)]: handleVxlanSpecialPacket
DEBU: 2016/02/29 02:46:34.192374 fastdp.send: odpFlowKeyFlowOp: InPortFlowKey{vport: 1}
DEBU: 2016/02/29 02:46:40.612722 ODP miss map[7:BlobFlowKey{type: 7, key: 0a2000010a2fffff01004000, mask: ffffffffffffffffffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: ff:ff:ff:ff:ff:ff} 11:BlobFlowKey{type: 11, key: 0800, mask: ffff} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 16:TunnelFlowKey{id: 000000000084dbb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 39603, tpdst: 6784} 3:InPortFlowKey{vport: 1} 20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff}] on port 1
DEBU: 2016/02/29 02:46:40.612807 getMissHandler: 1 0x57c9a0
DEBU: 2016/02/29 02:46:40.612840 vxlan missHandler: 0x57e8e0 82:43:51:a8:79:d4(host2) 1a:1e:21:54:f9:75(host1)
INFO: 2016/02/29 02:46:40.612878 Discovered remote MAC 82:43:51:a8:79:d4 at 82:43:51:a8:79:d4(host2)
DEBU: 2016/02/29 02:46:40.612903 bridge: {0 true} {82:43:51:a8:79:d4 ff:ff:ff:ff:ff:ff}
DEBU: 2016/02/29 02:46:40.613078 Creating new flow:
MultiFlowOp: false [
 MultiFlowOp: false [
  odpFlowKeyFlowOp: TunnelFlowKey{id: 000000000084dbb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11},
  MultiFlowOp: false [
   MultiFlowOp: false [
    odpFlowKeyFlowOp: EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: ff:ff:ff:ff:ff:ff},
    odpActionsFlowOp: [OutputAction{vport: 0}],
   ],
   DiscardingFlowOp,
  ],
 ],
 odpFlowKeyFlowOp: InPortFlowKey{vport: 1},
]
DEBU: 2016/02/29 02:46:40.613118 fastdp.send: odpFlowKeyFlowOp: TunnelFlowKey{id: 000000000084dbb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11}
DEBU: 2016/02/29 02:46:40.613142 fastdp.send: odpFlowKeyFlowOp: EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: ff:ff:ff:ff:ff:ff}
DEBU: 2016/02/29 02:46:40.613161 fastdp.send: odpActionsFlowOp: [OutputAction{vport: 0}]
DEBU: 2016/02/29 02:46:40.613176 fastdp.send: DiscardingFlowOp
DEBU: 2016/02/29 02:46:40.613187 fastdp.send: odpFlowKeyFlowOp: InPortFlowKey{vport: 1}
DEBU: 2016/02/29 02:46:41.612807 ODP miss map[20:BlobFlowKey{type: 20, key: 00000000, mask: ffffffff} 4:EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: ff:ff:ff:ff:ff:ff} 19:BlobFlowKey{type: 19, key: 00000000, mask: ffffffff} 16:TunnelFlowKey{id: 000000000084dbb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11, ttl: 64, tpsrc: 39603, tpdst: 6784} 15:BlobFlowKey{type: 15, key: 00000000, mask: ffffffff} 11:BlobFlowKey{type: 11, key: 0800, mask: ffff} 3:InPortFlowKey{vport: 1} 6:BlobFlowKey{type: 6, key: 0800, mask: ffff} 7:BlobFlowKey{type: 7, key: 0a2000010a2fffff01004000, mask: ffffffffffffffffffffffff} 2:BlobFlowKey{type: 2, key: 00000000, mask: ffffffff}] on port 1
DEBU: 2016/02/29 02:46:41.612882 getMissHandler: 1 0x57c9a0
DEBU: 2016/02/29 02:46:41.612910 vxlan missHandler: 0x57e8e0 82:43:51:a8:79:d4(host2) 1a:1e:21:54:f9:75(host1)
DEBU: 2016/02/29 02:46:41.612932 bridge: {0 true} {82:43:51:a8:79:d4 ff:ff:ff:ff:ff:ff}
DEBU: 2016/02/29 02:46:41.612985 Creating new flow:
MultiFlowOp: false [
 MultiFlowOp: false [
  odpFlowKeyFlowOp: TunnelFlowKey{id: 000000000084dbb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11},
  MultiFlowOp: false [
   MultiFlowOp: false [
    odpFlowKeyFlowOp: EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: ff:ff:ff:ff:ff:ff},
    odpActionsFlowOp: [OutputAction{vport: 0}],
   ],
   DiscardingFlowOp,
  ],
 ],
 odpFlowKeyFlowOp: InPortFlowKey{vport: 1},
]
DEBU: 2016/02/29 02:46:41.613009 fastdp.send: odpFlowKeyFlowOp: TunnelFlowKey{id: 000000000084dbb2, ipv4src: 192.168.48.12, ipv4dst: 192.168.48.11}
DEBU: 2016/02/29 02:46:41.613026 fastdp.send: odpFlowKeyFlowOp: EthernetFlowKey{src: 82:43:51:a8:79:d4, dst: ff:ff:ff:ff:ff:ff}
DEBU: 2016/02/29 02:46:41.613041 fastdp.send: odpActionsFlowOp: [OutputAction{vport: 0}]
DEBU: 2016/02/29 02:46:41.613054 fastdp.send: DiscardingFlowOp
DEBU: 2016/02/29 02:46:41.613071 fastdp.send: odpFlowKeyFlowOp: InPortFlowKey{vport: 1}
````

**Comment by dpw** [comment]
> @dpw can you comment on what is supposed to happen after the first broadcast/multicast packet?

What you might assume ought to happen - a rule gets created that handles subsequent packets.

This was definitely working at some point during fastdp development.  Too much time has passed for me to remember whether I tested it in the final 1.2.0 release.

From the logs, it looks like fastdp is trying to create an appropriate rule.  Obvious things to check:

* Did the correct rule really get created? `weave report` should be able to confirm this.

* So if the rule is there, is it definitely not firing?  `weave report` has the counters for this, IIRC.

**Comment by bboreham** [comment]
Thanks for the `weave report` tip.  Looks like not firing:

````
        "OverlayDiagnostics": {
            "fastdp": {
                "Vports": [
                    {
                        "ID": 0,
                        "Name": "weave",
                        "TypeName": "internal"
                    },
                    {
                        "ID": 1,
                        "Name": "vxlan-6784",
                        "TypeName": "vxlan"
                    }
                ],
                "Flows": [
                    {
                        "FlowKeys": [
                            "InPortFlowKey{vport: 0}",
                            "EthernetFlowKey{src: de:22:4e:20:60:05, dst: ff:ff:ff:ff:ff:ff}"
                        ],
                        "Actions": [
                            "SetTunnelAction{id: 000000000039e854, ipv4src: 192.168.48.11, ipv4dst: 192.168.48.12, ttl: 64, df: true}",
                            "OutputAction{vport: 1}"
                        ],
                        "Packets": 0,
                        "Bytes": 0,
                        "Used": 0
                    }
                ]
            },
            "sleeve": null
        },
````

**Comment by bboreham** [comment]
Shouldn't there be something about packets coming in over the tunnel, i.e. `InPortFlowKey vport: 1` ?  This is reported on the receiving side.

**Comment by bboreham** [comment]
If I repeat the experiment with a container attached on the sending and receiving side, i.e. not having done `weave expose`, then `weave report` shows no flows after a `ping 10.47.255.255` from inside the container, although the logs show it trying to create one.

**Comment by dpw** [comment]
> From the logs, it looks like fastdp is trying to create an appropriate rule.

I take that back - I confused the debugging output "Creating new flow" with "Creating ODP flow".  I think you added the former?  But the latter never appears.  Which means that no flows are being created.  Which seems very odd.

**Comment by bboreham** [comment]
Yes, I added that.  You can see all the changes on my branch at https://github.com/weaveworks/weave/compare/fastdp-flow-logging?expand=1

**Comment by bboreham** [comment]
It set `createFlow` to false when it encountered a `DiscardingFlowOp`, which in turn was inserted because `relayBroadcast()` found zero next hops.

**Comment by dpw** [comment]
`collectFlowOps` isn't accounting for `DiscardingFlowOp`.

**Comment by bboreham** [comment]
And, taking out https://github.com/weaveworks/weave/blob/master/router/network_router.go#L194-L196 makes the original problem go away.

**Comment by bboreham** [comment]
Maybe I'm confused about the meaning of `DiscardingFlowOp`.

**Comment by dpw** [comment]
> And, taking out https://github.com/weaveworks/weave/blob/master/router/network_router.go#L194-L196 makes the original problem go away.

Yeah, it would do.  Handling of `DiscardingFlowOp` is still wrong in `FlattenFlowOp` though, and thus in `FastDatapath send`.  It's supposed to simply mean "throw the packet away".

**Comment by bboreham** [comment]
We don't want to throw the packet away; we want to deliver it to every attached vport.

**Comment by rade** [comment]
>  It's supposed to simply mean "throw the packet away".

> We don't want to throw the packet away; we want to deliver it to every attached vport.

Is that the difference between `DiscardingFlowOp` and an "empty" `MultiFlowOp`?

**Comment by bboreham** [comment]
@rade I believe the code at https://github.com/weaveworks/weave/blob/master/router/fastdp.go#L283-L291 does the necessary; the trick is to not defeat that work elsewhere. An empty `MultiFlowOp` will achieve that trick, but I have also confirmed that `nil` will work.

**Comment by dpw** [comment]
> We don't want to throw the packet away; we want to deliver it to every attached vport.

There's no forwarding FlowOp for the packet.  When we consider forwarding in isolation, we are discarding the packet.

There is a injecting FlowOp for the packet.

The sum of a DiscardingFlowOp and some other FlowOp is that second FlowOp.

There might be a better name than DiscardingFlowOp (not NullFlowOp though, that doesn't communicate anything).

**Comment by dpw** [comment]
> Is that the difference between DiscardingFlowOp and an "empty" MultiFlowOp?

Behaviourally, nothing.

DiscardingFlowOps don't require an allocation, and it's used as the basis for `vetoFlowCreationFlowOp`.

**Comment by dpw** [comment]
> @rade I believe the code at https://github.com/weaveworks/weave/blob/master/router/fastdp.go#L283-L291 does the necessary; the trick is to not defeat that work elsewhere. An empty MultiFlowOp will achieve that trick

It fixes the bug you are concerned with, but leaves a latent bug.  The latent bug should be fixed, and that will fix the symptom you have run into.  I don't much mind whether `relayBroadcast` gets cleaned up or not.

> but I have also confirmed that nil will work.

Nack.  A nil FlowOp is not allowed, FlowOp methods get called unconditionally.  If you returned a nil FlowOp and it didn't blow up for you, you got lucky.

**Comment by dpw** [comment]
Yuck, it's even worse than I thought.

**Comment by dpw** [comment]
Who wrote that shitty code?

#2008 fixes things up in a minimal way.  The result still leaves something to be desired, but I'd need to look at the code more to have a good idea what (I can't remember if `odpFlowKeyFlowOp` and `vetoFlowCreationFlowOp` leak out to pcap).

**Comment by bboreham** [comment]
I don't see any references to FlowOps in `pcap.go` apart from the main `Pcap` struct that implements the interface.

**Comment by dpw** [comment]
> I don't see any references to FlowOps in pcap.go apart from the main Pcap struct that implements the interface.

In `sniff`.

**Comment by dpw** [comment]
(To elaborate: The whole point of `Discards()` is for pcap's `sniff`, to allow it to avoid a packet copy.  If those fastdp special FlowOps are never seen by pcap, then the result of their `Discards()` is moot.  So it would rule in or out some options for how to arrange things.)

**Comment by bboreham** [comment]
Fixed (minimally) by 27367ab292c58761266380add57c1c57a99a4f10 / #2008
